### ISSUE TITLE
regex with \p

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#1398]
Migrated from [rt.perl.org#2492](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=2492) (status was 'resolved')

Searchable as RT2492$

**Comment by p5pRT** [comment]
### From Richard.Foley@m.dasa.de

I know there is some mention of something along these lines in 'perldoc
perlre' and 'man utf8'\, but I've either misunderstood it \(&#8203;:\-\) or the
docs could say more\, or the error message could say more\, or \.\.\.&#8203;:

This code&#8203;:
\# \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
    my $str = 'abc\-\\pxyz';
&nbsp;    if \('xyz' =~ /$str/\) \{
&nbsp;            print "found '$str'\\n";
&nbsp;    \} 

Produces this lovely message&#8203;:
\# \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
Uncaught exception from user code&#8203;:
&nbsp;        Can't find unicode character property definition via main\->x or
x\.pl at unicode/Is/x\.pl line 0
&nbsp;        Carp&#8203;::croak\('Can\\'t find unicode character property definition
via main\->x or\.\.\.'\) called at /usr/local/lib/perl5/5\.6\.0/utf8\_heavy\.pl
line 8
&nbsp;        utf8&#8203;::croak\('Can\\'t find unicode character property definition
via main\->x or\.\.\.'\) called at /usr/local/lib/perl5/5\.6\.0/utf8\_heavy\.pl
line 30
&nbsp;        utf8&#8203;::SWASHNEW\('utf8'\, 'x'\, ''\, 0\, 0\, 0\) called at
/usr/local/lib/perl5/5\.6\.0/utf8\_heavy\.pl line 78
&nbsp;        utf8&#8203;::SWASHNEW\('utf8'\, ''\, '\# comment^J\+utf8&#8203;::x^J'\, 1\, 0\) called
at unicode/Is/x\.pl line 0  

If on the other hand we use 'abc\-\\xyz' \(\\x instead of \\p\) we get some
hint as to what may be going on&#8203;:
\# \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\- 
Illegal hexadecimal digit 'y' ignored at \./trial line 6 \(\#1\)
 
&nbsp;    \(W digit\) You may have tried to use a character other than 0 \- 9 or
A \- F\, a \- f
&nbsp;    in a hexadecimal number\.  Interpretation of the hexadecimal number
stopped
&nbsp;    before the illegal character\.
 
Segmentation fault  

BTW&#8203;: No amount of saying 'use utf8' or 'no utf8' makes any difference
either\.

Ciao
Richard Foley


---
richard@&#8203;rfi\.net
'Ciao' \- shorter than 'Aufwiedersehen'

**Comment by p5pRT** [comment]
### From @gsar

On Tue\, 21 Mar 2000 15&#8203;:29&#8203;:17 \+0100\, Richard Foley wrote&#8203;:
>I know there is some mention of something along these lines in 'perldoc
>perlre' and 'man utf8'\, but I've either misunderstood it \(&#8203;:\-\) or the
>docs could say more\, or the error message could say more\, or \.\.\.&#8203;:
>
>This code&#8203;:
>\# \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
>^C    my $str = 'abc\-\\pxyz';
>    if \('xyz' =~ /$str/\) \{

You need&#8203;:

&nbsp;        if \('xyz' =~ /\\Q$str/\) \{


Sarathy
gsar@&#8203;ActiveState\.com

**Comment by p5pRT** [comment]
### From @vanstyn

In \<38D7873D\.62877924@&#8203;m\.dasa\.de>\, Richard Foley writes&#8203;:
:\# \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
:     my $str = 'abc\-\\pxyz';
:    if \('xyz' =~ /$str/\) \{
:            print "found '$str'\\n";
:    \} 
:
\[\.\.\.\]
:If on the other hand we use 'abc\-\\xyz' \(\\x instead of \\p\) we get some
:hint as to what may be going on&#8203;:
:\# \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\- 
:Illegal hexadecimal digit 'y' ignored at \./trial line 6 \(\#1\)
: 
:    \(W digit\) You may have tried to use a character other than 0 \- 9 or
:A \- F\, a \- f
:    in a hexadecimal number\.  Interpretation of the hexadecimal number
:stopped
:    before the illegal character\.
: 
:Segmentation fault  

\(Not mentioned\, but this currently coredumps for me with RC3\.\)

This occurs because diagnostics\.pm\, within a $SIG\{\_\_WARN\_\_\} handler\,
uses regexps; the warning is generated within a regexp\, so this attempts
to use regexps reentrantly\. On this occasion it happens to corrupt one
of the regexp statics to point to PL\_diehook\, which is then overwritten;
the coredump occurs when an attempt is made to treat the overwritten
PL\_diehook as an sv\.

Possible fixes&#8203;: 1\) change diagnostics\.pm to croak on the first warning;
2\) change diagnostics\.pm to stack all warnings and process only when a
fatal is seen\, or when the program ends; 3\) make regexps reentrant\.

The most recent version I have that does not coredump on this is \_63\,
but I don't understand why not \- maybe just lucky\.

Hugo

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

Gurusamy Sarathy wrote&#8203;:
> 
> On Tue\, 21 Mar 2000 15&#8203;:29&#8203;:17 \+0100\, Richard Foley wrote&#8203;:
> >I know there is some mention of something along these lines in 'perldoc
> >perlre' and 'man utf8'\, but I've either misunderstood it \(&#8203;:\-\) or the
> >docs could say more\, or the error message could say more\, or \.\.\.&#8203;:
> >
> >This code&#8203;:
> >\# \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> >^C     my $str = 'abc\-\\pxyz';
> >       if \('xyz' =~ /$str/\) \{
> 
> You need&#8203;:
> 
>         if \('xyz' =~ /\\Q$str/\) \{
Surely that quotes the backslash as well?

Which is not what I want\, while the $str is built elsewhere and may have
it's own regex which I don't want to have further interpreted by \\Q\.\.\.\\E

Ciao
Richard Foley


---
richard@&#8203;rfi\.net
'Ciao' \- shorter than 'Aufwiedersehen'

**Comment by p5pRT** [comment]
### From @gsar

On Tue\, 21 Mar 2000 18&#8203;:19&#8203;:40 GMT\, Hugo wrote&#8203;:
>&#8203;:If on the other hand we use 'abc\-\\xyz' \(\\x instead of \\p\) we get some
>&#8203;:hint as to what may be going on&#8203;:
>&#8203;:\# \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\- 
>&#8203;:Illegal hexadecimal digit 'y' ignored at \./trial line 6 \(\#1\)
>&#8203;: 
>&#8203;:    \(W digit\) You may have tried to use a character other than 0 \- 9 or
>&#8203;:A \- F\, a \- f
>&#8203;:    in a hexadecimal number\.  Interpretation of the hexadecimal number
>&#8203;:stopped
>&#8203;:    before the illegal character\.
>&#8203;: 
>&#8203;:Segmentation fault  
>
>\(Not mentioned\, but this currently coredumps for me with RC3\.\)
>
>This occurs because diagnostics\.pm\, within a $SIG\{\_\_WARN\_\_\} handler\,
>uses regexps; the warning is generated within a regexp\, so this attempts
>to use regexps reentrantly\. On this occasion it happens to corrupt one
>of the regexp statics to point to PL\_diehook\, which is then overwritten;
>the coredump occurs when an attempt is made to treat the overwritten
>PL\_diehook as an sv\.

Thanks for the analysis\.

>Possible fixes&#8203;: 1\) change diagnostics\.pm to croak on the first warning;
>2\) change diagnostics\.pm to stack all warnings and process only when a
>fatal is seen\, or when the program ends; 3\) make regexps reentrant\.

Here's a fourth alternative that should be a good enough medium\-term
fix\.  I also found that some of the new calls to the hooks aren't setting
up a new stacklevel like they should\.

This has been stress\-tested with Purify\.


Sarathy
gsar@&#8203;ActiveState\.com
<details><summary>Inline Patch</summary>

```diff
-----------------------------------8<-----------------------------------
Change 5877 by gsar@auger on 2000/03/22 15:38:49

	save RE context when calling __WARN__ and __DIE__ hooks;
	add missing stack-of-stacks logic for new code that came
	in with lexical warnings (this fixes coredumps from
	-Mdiagnostics when the warnings happens to come from within
	the RE engine)

Affected files ...

... //depot/perl/util.c#180 edit

Differences ...

==== //depot/perl/util.c#180 (text) ====
Index: perl/util.c
--- perl/util.c.~1~	Wed Mar 22 07:38:52 2000
+++ perl/util.c	Wed Mar 22 07:38:52 2000
@@ -1515,6 +1515,7 @@
 	    SV *msg;
 
 	    ENTER;
+	    save_re_context();
 	    if (message) {
 		msg = newSVpvn(message, msglen);
 		SvREADONLY_on(msg);
@@ -1604,6 +1605,7 @@
 	    SV *msg;
 
 	    ENTER;
+	    save_re_context();
 	    msg = newSVpvn(message, msglen);
 	    SvREADONLY_on(msg);
 	    SAVEFREESV(msg);
@@ -1697,6 +1699,7 @@
 	    SV *msg;
 
 	    ENTER;
+	    save_re_context();
 	    msg = newSVpvn(message, msglen);
 	    SvREADONLY_on(msg);
 	    SAVEFREESV(msg);
@@ -1810,15 +1813,17 @@
                 SV *msg;
  
                 ENTER;
+		save_re_context();
                 msg = newSVpvn(message, msglen);
                 SvREADONLY_on(msg);
                 SAVEFREESV(msg);
  
+		PUSHSTACKi(PERLSI_DIEHOOK);
                 PUSHMARK(sp);
                 XPUSHs(msg);
                 PUTBACK;
                 call_sv((SV*)cv, G_DISCARD);
- 
+		POPSTACK;
                 LEAVE;
             }
         }
@@ -1843,21 +1848,23 @@
             SAVESPTR(PL_warnhook);
             PL_warnhook = Nullsv;
             cv = sv_2cv(oldwarnhook, &stash, &gv, 0);
-                LEAVE;
+	    LEAVE;
             if (cv && !CvDEPTH(cv) && (CvROOT(cv) || CvXSUB(cv))) {
                 dSP;
                 SV *msg;
  
                 ENTER;
+		save_re_context();
                 msg = newSVpvn(message, msglen);
                 SvREADONLY_on(msg);
                 SAVEFREESV(msg);
  
+		PUSHSTACKi(PERLSI_WARNHOOK);
                 PUSHMARK(sp);
                 XPUSHs(msg);
                 PUTBACK;
                 call_sv((SV*)cv, G_DISCARD);
- 
+		POPSTACK;
                 LEAVE;
                 return;
             }
End of Patch.
```
</details>
